coderd/unhanger/detector.go

```diff
@@ -0,0 +1,318 @@
+func New(ctx context.Context, db database.Store, pubsub database.Pubsub, log slog.Logger, tick <-chan time.Time) *Detector {
+	le := &Detector{
+		//nolint:gocritic // Hang detector has a limited set of permissions.
+		ctx:    dbauthz.AsHangDetector(ctx),
+		db:     db,
+		pubsub: pubsub,
+		tick:   tick,
+		log:    log,
+		done:   make(chan struct{}),
+	}
+	return le
+}
+
+// WithStatsChannel will cause Executor to push a RunStats to ch after
+// every tick. This push is blocking, so if ch is not read, the detector will
+// hang.
+func (d *Detector) WithStatsChannel(ch chan<- Stats) *Detector {
+	d.stats = ch
+	return d
+}
+
+// Run will cause the detector to detect and unhang provisioner jobs on every
+// tick from its channel. It will stop when its context is Done, or when its
+// channel is closed.
+//
+// Run should only be called once.
+func (d *Detector) Run() {
+	go func() {
+		defer close(d.done)
+		for {
+			select {
+			case <-d.ctx.Done():
+				return
+			case t, ok := <-d.tick:
+				if !ok {
+					return
+				}
+				stats := d.runOnce(t)
+				if stats.Error != nil && !xerrors.As(stats.Error, &acquireLockError{}) {
+					d.log.Warn(d.ctx, "error running workspace build hang detector once", slog.Error(stats.Error))
+				}
+				if len(stats.HungJobIDs) != 0 {
+					d.log.Warn(d.ctx, "detected (and terminated) hung provisioner jobs", slog.F("job_ids", stats.HungJobIDs))
+				}
+				if d.stats != nil {
+					select {
+					case <-d.ctx.Done():
+						return
+					case d.stats <- stats:
+					}
+				}
+			}
+		}
+	}()
+}
+
+// Wait will block until the detector is stopped.
+func (d *Detector) Wait() {
+	<-d.done
+}
+
+func (d *Detector) runOnce(t time.Time) Stats {
+	ctx, cancel := context.WithTimeout(d.ctx, 5*time.Minute)
+	defer cancel()
+
+	stats := Stats{
+		HungJobIDs: []uuid.UUID{},
+		Error:      nil,
+	}
+
+	err := d.db.InTx(func(db database.Store) error {
+		err := db.AcquireLock(ctx, database.LockIDHangDetector)
+		if err != nil {
+			// If we can't acquire the lock, it means another instance of the
+			// hang detector is already running in another coder replica.
+			// There's no point in waiting to run it again, so we'll just retry
+			// on the next tick.
+			d.log.Info(ctx, "skipping workspace build hang detector run due to lock", slog.Error(err))
+			// This error is ignored.
+			return &acquireLockError{err: err}
+		}
+		d.log.Info(ctx, "running workspace build hang detector")
+
+		// Find all provisioner jobs that are currently running but have not
+		// received an update in the last 5 minutes.
+		jobs, err := db.GetHungProvisionerJobs(ctx, t.Add(-HungJobDuration))
+		if err != nil {
+			return xerrors.Errorf("get hung provisioner jobs: %w", err)
+		}
+
+		// We only use errgroup here for convenience of API, not for early
+		// cancellation. This means we only return nil errors in th eg.Go.
+		eg := errgroup.Group{}
+		// Limit the concurrency to avoid overloading the database.
+		eg.SetLimit(10)
```

Comment:
This appears unused? It probably doesn't make sense to use it (due to transaction) either.